### PR TITLE
[otp/rom] Add entropy_src health OTP settings.

### LINKED ISSUE #9492: [rom] Populate entropy_src health test settings
The following is the high level boot flow regarding the entropy complex (courtesy of @moidx).

**High Level Boot Flow - Software** 

- ROM
   - EDN0 in BOOT_REQ_MODE to enable downstream consumers (SRAM scrambling, et al). Done in assembly.
   - Cheap & fast entropy required by software for countermeasures against fault attacks.
   - Periodic expectation checks for critical registers. This may include entropy complex configuration registers.
- ROM_EXT
   - Cheap & fast entropy required by software for countermeasures against attacks.
   - If attestation or ownership transfer (this is a conditional slow boot path):
      - FIPS: entropy_src.conditioner KAT test. (possibly using the FW OVR and ENTOPY_DATA interface).
      - FIPS: csrng KAT test, utilizing the sw instance (e.g. #7505)
      - FIPS: enable entropy_src in FIPS mode, GetHealthTest results.
      - FIPS: crypto operations: csrng software instance and/or EDN1.
   - Periodic expectation checks for critical registers. This may include entropy complex configuration registers. 
- Silicon Owner software (BL0 or Kernel):  Same as ROM_EXT

For the cheap / fast entropy in ROM, there needs to be some additional configuration of entropy_src health test settings.
These should be read out from OTP and programmed into entropy_src.

We will have to define a simple / light-weight scheme to check there is data in OTP and copy over, done in assembly.
When the C-runtime is initialized, we then need to do consistency checks to make sure what's programmed matches what's in OTP.

### PR DESCRIPTION
**Pull request by moidx**
This change adds OTP settings for the entropy_src health tests. These settings must be loaded into OTP before enabling ROM execution otherwise the entropy source will trigger a health test alert.

fixes #9492

### CODE REVIEW

**Comment by moidx** [comment]
@alphan this change unconditionally loads words from OTP before enabling the entropy source. I am not sure if it makes sense to also add a separate configuration word to enable/disable all these writes. If we keep it as is, it may still be possible for the ROM to boot when with uninitialized OTP settings, but the entropy health check alert would have to be disabled. I will do some testing in parallel to check this.

File: hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson [hunk 1]
```diff
@@ -184,6 +184,46 @@
                     name: "CREATOR_SW_CFG_RMA_SPIN_CYCLES",
                     size: "4"
                 }
+                {
+                    name: "CREATOR_SW_RNG_REPCNT_THRESHOLDS",
+                    size: "4"
```
Comment (moidx) [inline hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson:189]: For all these settings, I am allocating enough bits to set both boot and continuous entropy mode. ROM_EXT can then relax the continuous mode settings if needed.

File: hw/ip/otp_ctrl/data/BUILD [hunk 2]
```diff
@@ -78,6 +78,18 @@ otp_json(
...
+                "CREATOR_SW_RNG_REPCNT_THRESHOLDS": "0xffffffff",
+                "CREATOR_SW_RNG_REPCNTS_THRESHOLDS": "0xffffffff",
+                "CREATOR_SW_RNG_ADAPTP_HI_THRESHOLDS": "0xffffffff",
+                "CREATOR_SW_RNG_ADAPTP_LO_THRESHOLDS": "0x0",
+                "CREATOR_SW_RNG_BUCKET_THRESHOLDS": "0xffffffff",
+                "CREATOR_SW_RNG_MARKOV_HI_THRESHOLDS": "0xffffffff",
+                "CREATOR_SW_RNG_MARKOV_LO_THRESHOLDS": "0x0",
+                "CREATOR_SW_RNG_EXTHT_HI_THRESHOLDS": "0xffffffff",
+                "CREATOR_SW_RNG_EXTHT_LO_THRESHOLDS": "0x0",
+                "CREATOR_SW_RNG_ALERT_THRESHOLD": "0xfffd0002",
```
Comment (alphan) [inline hw/ip/otp_ctrl/data/BUILD:92]: nit: I think all other items start with `CREATOR_SW_CFG`. Would you mind adding `_CFG` to these as well? If you think the names get too long we can shorten THRESHOLD as THOLD.

File: hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson (hunk 1 above)
Comment (alphan) [reply hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson:189]: I assume these are two `uint16_t`s packed into a single `uint32_t`, correct?

File: hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson [hunk 3]
```diff
@@ -184,6 +184,46 @@
...
+                }
+                {
+                    name: "CREATOR_SW_RNG_REPCNTS_THRESHOLDS",
```
Comment (alphan) [inline hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson:192]: nit: I'm not very familiar with entropy_src but just wanted to confirm: Is the plurality of REPCNT the only difference between this item and the one above?

File: sw/device/silicon_creator/rom/rom_start.S [hunk 4]
```diff
@@ -302,6 +302,40 @@ LABEL_FOR_TEST(kRomStartWatchdogEnabled)
   // The following sequence enables the minimum level of entropy required to
   // initialize memory scrambling, as well as the entropy distribution network.
   li a0, TOP_EARLGREY_ENTROPY_SRC_BASE_ADDR
+  li a1, (TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR + \
+            OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET)
+
+  // Load entropy src health check settings from OTP.
```
Comment (alphan) [inline sw/device/silicon_creator/rom/rom_start.S:308]: nit: We can consider using `crt_section_copy` if the registers are consecutive. We can then confirm that everything is as expected in rom.c. I think we should do it regardless to harden this configuration.

**Comment by alphan** [comment]
> @alphan this change unconditionally loads words from OTP before enabling the entropy source. I am not sure if it makes sense to also add a separate configuration word to enable/disable all these writes. If we keep it as is, it may still be possible for the ROM to boot when with uninitialized OTP settings, but the entropy health check alert would have to be disabled. I will do some testing in parallel to check this.

I was wondering about the same thing, thanks for pointing this out @moidx. We skip alert configuration in TEST_UNLOCKED* so we should be able to boot with default OTP values as long as entropy_src is ok with all zeroes. This definitely introduces a dependency between these OTP items that I should capture in the manufacturing doc.

**Comment by alphan** [comment]
This reminds me: we should probably add a simple boot test with all default OTP values in TEST_UNLOCKED.

**Review by alphan — APPROVED** [review]

File: hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson (hunk 3 above)
Comment (moidx) [reply hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson:192]: That is correct. `REPCNTS` is a different configuration compared to `REPCNT`. Single bit versus symbol based repetition count tests. I believe symbol bases does it in an N-bit granularity based on how many RNG bitstreams are provided by AST.

Comment (alphan) [reply hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson:192]: Thanks for the explanation @moidx!

File: hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson (hunk 1 above)
Comment (moidx) [reply hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson:189]: This is correct. The manufacturing script will have to be aware of this.

File: hw/ip/otp_ctrl/data/BUILD (hunk 2 above)
Comment (moidx) [reply hw/ip/otp_ctrl/data/BUILD]: Done.

File: sw/device/silicon_creator/rom/rom_start.S (hunk 4 above)
Comment (moidx) [reply sw/device/silicon_creator/rom/rom_start.S]: Done. I added a second commit implementing the check.

**Comment by moidx** [comment]
Hi @alphan, can you take another look? I ended adding the ROM checks to this PR. You can focus on the 2nd commit if you are satisfied with the resolution of the previous comments. Thanks!

**Comment by moidx** [comment]
Hi @alphan, gentle ping.

**Comment by alphan** [comment]
requested a re-run since the log is not available anymore (at least for cw310 rom tests)

File: sw/device/silicon_creator/rom/rom_start.S [hunk 5]
```diff
@@ -304,9 +304,28 @@ LABEL_FOR_TEST(kRomStartWatchdogEnabled)
...
+            ENTROPY_SRC_EXTHT_LO_THRESHOLDS_REG_OFFSET + 4)
+  li   a2, (TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR + \
+            OTP_CTRL_SW_CFG_WINDOW_REG_OFFSET + \
+            OTP_CTRL_PARAM_CREATOR_SW_CFG_RNG_REPCNT_THRESHOLDS_OFFSET)
+  call crt_section_copy
+
+
+  li a0, TOP_EARLGREY_ENTROPY_SRC_BASE_ADDR
+
+  // Configure the entropy source health alert threshold.
```
Comment (alphan) [inline sw/device/silicon_creator/rom/rom_start.S:320]: nit:
```suggestion
  // Configure the entropy source health alert threshold.
  li a0, TOP_EARLGREY_ENTROPY_SRC_BASE_ADDR
```

File: sw/device/silicon_creator/lib/drivers/rnd.c [hunk 6]
```diff
@@ -1,30 +1,113 @@
...
+  kNumHealthRegisters = 9,
 };
 
+// Check the number of health registers covered by this driver.
+static_assert(kNumHealthRegisters ==
+                  (ENTROPY_SRC_EXTHT_LO_THRESHOLDS_REG_OFFSET -
+                   ENTROPY_SRC_REPCNT_THRESHOLDS_REG_OFFSET) /
+                          sizeof(uint32_t) +
+                      1,
+              "foobar");
```
Comment (alphan) [inline sw/device/silicon_creator/lib/drivers/rnd.c:33]: nit: can you update the `static_assert` message?

File: sw/device/silicon_creator/lib/error.h [hunk 7]
```diff
@@ -126,7 +127,9 @@ enum module_ {
   \
   X(kErrorSpiDevicePayloadOverflow,   ERROR_(1, kModuleSpiDevice, kInternal)), \
   \
-  X(kErrorAstInitNotDone,             ERROR_(1, kModuleAst, kInternal))
+  X(kErrorAstInitNotDone,             ERROR_(1, kModuleAst, kInternal)), \
+  \
+  X(kErrorRndBadCrc32,                ERROR_(5, KModuleRnd, kInvalidArgument))
```
Comment (alphan) [inline sw/device/silicon_creator/lib/error.h:132]: nit: shouldn't this start from 1?
```suggestion
  X(kErrorRndBadCrc32,                ERROR_(1, KModuleRnd, kInvalidArgument))
```

**Comment by alphan** [comment]
@moidx 👋🏼

File: sw/device/silicon_creator/rom/rom_start.S (hunk 5 above)
Comment (moidx) [reply sw/device/silicon_creator/rom/rom_start.S]: Done

File: sw/device/silicon_creator/lib/drivers/rnd.c (hunk 6 above)
Comment (moidx) [reply sw/device/silicon_creator/lib/drivers/rnd.c]: Done

File: sw/device/silicon_creator/lib/error.h (hunk 7 above)
Comment (moidx) [reply sw/device/silicon_creator/lib/error.h]: Done

**Comment by moidx** [comment]
> Thanks @moidx and apologies for the delay. Just some nits. Also, `CREATOR_SW_CFG_RNG_EN` is currently used for gating entropy usage in countermeasures. Do you think we should gate the config in rom_start.S and check in the rnd driver with it or are we confident that the test & alert mechanism will work fine?

Hi @alphan, I think we are ok here, but will add a top-level test to confirm the behavior.

**Comment by alphan** [comment]
> Hi @alphan, I think we are ok here but will add a top-level test to confirm the behavior.

Thanks @moidx! Just one failure in cw310 tests:

```
//sw/device/silicon_creator/rom/e2e:rom_e2e_default_otp_bootup_fpga_cw310_rom FAILED in 2 out of 2 in 16.7s
```

I guess the OTP image that this test uses needs to be updated.

**Pull request merged**